agent/agent.go

```diff
@@ -265,20 +277,102 @@ func (a *Agent) loop(ctx context.Context) error {
 	if a.opts.MaxAgentSleepInterval != 0 {
 		maxAgentSleepInterval = a.opts.MaxAgentSleepInterval
 	}
-	agentSleepInterval := minAgentSleepInterval
+	taskCtx, cancel := context.WithCancel(ctx)
 
-	var jitteredSleep time.Duration
-	tskCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	lc := &loop{
+		tc:                    &taskContext{},
+		needPostGroup:         false,
+		taskCtx:               taskCtx,
+		cancel:                cancel,
+		timer:                 time.NewTimer(0),
+		minAgentSleepInterval: minAgentSleepInterval,
+		maxAgentSleepInterval: maxAgentSleepInterval,
+		agentSleepInterval:    minAgentSleepInterval,
+	}
+
+	return lc
+}
+
+func (a *Agent) processNextTask(ctx context.Context, l *loop, nextTask *apimodels.NextTaskResponse) (shouldContinue bool, shouldExit bool) {
+	if nextTask.ShouldExit {
+		grip.Notice("Next task response indicates agent should exit.")
+		shouldExit = true
+		return
+	}
+	// if the host's current task group is finished we teardown
+	if nextTask.ShouldTeardownGroup {
+		a.runPostGroupCommands(ctx, l.tc)
+		l.needPostGroup = false
+		// Running the post group commands implies exiting the group, so
+		// destroy prior task information.
+		l.tc = &taskContext{}
+		l.timer.Reset(0)
+		l.agentSleepInterval = l.minAgentSleepInterval
+		shouldContinue = true
+		return
+	}
+	if nextTask.TaskId != "" {
+		if nextTask.TaskSecret == "" {
+			grip.Critical(message.Fields{
+				"message": "task response missing secret",
+				"task":    l.tc.task.ID,
+			})
+			l.timer.Reset(0)
+			l.agentSleepInterval = l.minAgentSleepInterval
+			shouldContinue = true
+			return
+		}
+		prevLogger := l.tc.logger
+		l.tc = a.prepareNextTask(ctx, nextTask, l.tc)
+		if prevLogger != nil {
+			grip.Error(errors.Wrap(prevLogger.Close(), "closing the previous logger producer"))
+		}
+
+		grip.Error(message.WrapError(a.fetchProjectConfig(ctx, l.tc), message.Fields{
+			"message": "error fetching project config; will attempt at a later point",
+			"task":    l.tc.task.ID,
+		}))
+
+		a.jasper.Clear(ctx)
+		l.tc.jasper = a.jasper
+		taskShouldExit, err := a.runTask(l.taskCtx, l.tc)
+		shouldExit = taskShouldExit
+		if err != nil {
+			grip.Critical(message.WrapError(err, message.Fields{
+				"message": "error running task",
+				"task":    l.tc.task.ID,
+			}))
+			l.timer.Reset(0)
+			l.agentSleepInterval = l.minAgentSleepInterval
+			shouldContinue = true
+			return
+		}
+		if shouldExit {
+			return
+		}
+		l.needPostGroup = true
+		l.timer.Reset(0)
+		l.agentSleepInterval = l.minAgentSleepInterval
+		shouldContinue = true
+		return
+	} else if l.needPostGroup {
+		a.runPostGroupCommands(ctx, l.tc)
+		l.needPostGroup = false
+		// Running the post group commands implies exiting the group, so
+		// destroy prior task information.
+		l.tc = &taskContext{}
+	}
+	return
+}
 
-	timer := time.NewTimer(0)
-	defer timer.Stop()
+func (a *Agent) loop(ctx context.Context) error {
```

Comment:
I actually don't see this loop as much cleaner than how it was written before - the original desire was to take the complex part (the `processNextTask` logic) and turn that into a helper function, not the loop itself. This change makes things more stateful than they were before, because now you have to track the `loop` parameters as the agent iterates.